### PR TITLE
Working example

### PR DESCRIPTION
**Pull request by nicomness**
Just some working examples, not intended to be merged into main.

### CODE REVIEW

File: src/main/java/com/einabit/client/EinabitClient.java [hunk 1]
```diff
@@ -84,7 +93,7 @@ public void tap(final String variable, final EinabitServerListener callback) {
             int readBytes;
 
             byte[] buffer = new byte[BUFFER_SIZE];
-            while ((readBytes = dataInputStream.read(buffer)) != -1) {
+            while ((readBytes = dataInputStream.read(buffer)) != -1 && !Thread.currentThread().isInterrupted()) {
```
Comment (k1r0s) [inline src/main/java/com/einabit/client/EinabitClient.java:101]: Can you please take a look on this?

https://github.com/Einabit/client.java/pull/10/files#diff-2596d8b447cd46c7d81efc77d9439d94c2fda5515c77db849bbe0a959923676eR81-R94

We need to run the while loop until the thread is interrupted, the server or the datainputstream shouldn't drive client behavior. The datainputstream may be empty for a period of time if the server does not find any changes to inform.

thoughts?

**Comment by k1r0s** [comment]
plz test out last changes as encryption does not work. Didn't check deeply

**Comment by k1r0s** [comment]
Please check this piece of code as it does not work as expected

Running the thread within completablefuture does nothing

```java
import com.einabit.client.EinabitClient;
import com.einabit.client.EinabitServerListener;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

// import java.util.concurrent.CompletableFuture;

public class Main {

  private static final String ENV_VARIABLES = "VARIABLES";
  private static final String ENV_API_HOST = "API_HOST";
  private static final String ENV_DB_HOST = "DB_HOST";
  private static final String ENV_DB_USER = "DB_USER";
  private static final String ENV_DB_PASS = "DB_PASS";
  private static final String ENV_DB_NAME = "DB_NAME";

  public static void main(String[] args) {

    System.out.println("Starting app");

    String url = "jdbc:postgresql://" + System.getenv(ENV_DB_HOST) + ":5432/" + System.getenv(ENV_DB_NAME);
    final EinabitClient client = EinabitClient.builder().host(System.getenv(ENV_API_HOST)).build();

    try {
      final Connection cn = DriverManager.getConnection(url, System.getenv(ENV_DB_PASS), System.getenv(ENV_DB_PASS));

      System.out.println("Looking for variables: " + System.getenv(ENV_VARIABLES));
      for (String auxvarname: System.getenv(ENV_VARIABLES).split(",")) {

        final Thread myCustomRunnable = new Thread(new MyRunnable(client, auxvarname, (String auxval) -> {
          String[] auxsplit = auxval.trim().split(",");

          try {
            Statement statement = cn.createStatement();

            String sql = "INSERT INTO iot_registry (name, ts, value) VALUES ('$1', $2, $3)";

            sql = sql.replace("$1", auxvarname);
            sql = sql.replace("$2", auxsplit[0]);
            sql = sql.replace("$3", auxsplit[1]);

            System.out.println("Executed sql statement: " + sql);
            statement.execute(sql);

          } catch (SQLException e) {
            e.printStackTrace();
          }

        }), "my-runnable-thread");

        System.out.println("Subscribed for changes on " + auxvarname);

        myCustomRunnable.start(); // this works
        // CompletableFuture.runAsync(myCustomRunnable::start); // this does not work :|
      }

    } catch (SQLException e) {
      e.printStackTrace();
    }
  }

  private record MyRunnable(EinabitClient client, String varname, EinabitServerListener listener) implements Runnable {

    @Override
    public void run() {
      System.out.println("Client.tap for " + varname);
      client.tap(varname, listener);
    }
  }
}

```

**Comment by k1r0s** [comment]
this is the above setup

```yml
version: "3.9"
services:
  application:
    build:
      context: app/
      dockerfile: Dockerfile.app
    environment:
      VARIABLES: crg1,temp1
      API_HOST: sandbox
      DB_HOST: postgres
      DB_USER: docker
      DB_PASS: docker
      DB_NAME: docker
    depends_on:
      sandbox:
        condition: service_started
      postgres:
        condition: service_healthy
  sandbox:
    image: einabit/sandbox
    environment:
      CONFIG: |
        crg1:
          value:
            min: 20
            max: 40
            step: 2
          interval:
            ms: 1200
            skip: .5
        temp1:
          value:
            min: 1
            max: 50
            step: 1
          interval:
            ms: 200
            skip: .3
  postgres:
    image: postgres
    environment:
      POSTGRES_USER: root
      POSTGRES_PASSWORD: 1234
      APP_DB_USER: docker
      APP_DB_PASS: docker
      APP_DB_NAME: docker
    volumes:
      - ./db:/docker-entrypoint-initdb.d/
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 5s
      timeout: 5s
      retries: 5
    ports:
      - 5432:5432

```

**Comment by nicomness** [comment]
We have to keep in mind a couple of things here:

1. When you use a completable future you have more control over the executor and therefore over the thread pool (in this case, it's just using a default one), but all of that comes at a cost... it requires a bit of time to create a thread pool and start scheduling those threads. 

2. The main thread is the parent and it runs in the main thread pool.

Being said that what's going on here is that the main thread terminates before the other pool is created so it never schedules those threads. We have to be very careful in cases where we are doing nothing in the main thread and we must avoid it at all costs since it will lead to wasted CPU resources apart from unwanted behaviors. It's better to run something in the main thread in addition to n threads (which can be run on the same thread pool or a different one).

**Pull request closed without merging**